FIAT/finite_element.py

```diff
@@ -27,6 +28,82 @@
 
 
 class FiniteElement(object):
+    """Class implementing a basic abstraction template for special
+    types of finite elements, like traces. Most special methods will
+    need to be implemented by the special element in its module."""
+
+    def __init__(self, ref_el, order, formdegree=None, mapping="affine"):
+
+        self.order = order
+        self.formdegree = formdegree
+        self.mapping = mapping
+
+        self.ref_el = ref_el
+
+    def degree(self):
+        "Return the degree of the (embedding) polynomial space."
+        raise NotImplementedError()
```

Comment:
This needs some more thought because having a bunch of methods raising `NotImplementedError` is not useful. I assume `CiarletElement` is what `FiniteElement` had previously been, so what we need is an idea for what the new `FiniteElement` should be. I guess this should the bare minimum that we need to use a finite element in Firedrake (?)

Any way, if you want to list a bunch of methods that the derived classes should (must) implemented, [abc](https://docs.python.org/2/library/abc.html) is much better idea than raising empty `NonImplementedError`s.